tests/test_cli.py

```diff
@@ -0,0 +1,140 @@
+    {'add_tags': ['-', 'foo', 'baz', 'baz'], 'tags_except': ',foo baz baz,', 'immutable': False},
+    {'tag': ['-', 'foo,', ',baz,', ',baz'], 'tags_except': ',baz,foo,', 'immutable': True},
+    {'add_tags': ['foo,baz,bar'], 'tag': ['baz,qux'],
+     'tags_fetch': False, 'tags_in': ',bar,baz,foo,qux,'},
+    {'add_tags': ['+', 'foo,baz,bar'], 'tag': ['baz,', 'qux,'],
+     'tags_fetch': False, 'tags_in': ',bar,baz,foo,qux,'},
+    {'add_tags': ['foo,baz,', 'bar,'], 'tag': ['+', 'baz,qux'],
+     'tags_fetch': False, 'tags_in': ',bar,baz,foo,qux,'},
+    {'add_tags': ['-', 'foo,baz,bar,'], 'tag': ['baz,', 'qux,'],
+     'tags_fetch': False, 'tags_in': ',baz,qux,', 'tags_except': ',bar,baz,foo,'},
+    {'add_tags': ['foo,baz,', 'bar,'], 'tag': ['-', 'baz,qux'],
+     'tags_fetch': False, 'tags_in': ',bar,baz,foo,', 'tags_except': ',baz,qux,'},
+    {'add_tags': ['-', 'foo,baz,bar,'], 'tag': ['-', 'baz,', 'qux,'], 'tags_except': ',bar,baz,foo,qux,'},
+    {'add_tags': ['-', 'foo,baz,', 'bar,'], 'tag': ['+', 'baz,', 'qux,'],
+     'tags_in': ',baz,qux,', 'tags_except': ',bar,baz,foo,'},
+    {'add_tags': ['+', 'foo,baz,', 'bar,'], 'tag': ['-', 'baz,', 'qux,'],
+     'tags_in': ',bar,baz,foo,', 'tags_except': ',baz,qux,'},
+])
+def test_add(stdin, bdb, prompt, value_params, fetch_params):
+    _test_add(bdb, prompt, **value_params, **fetch_params)
+
+def _test_add(bdb, prompt, *, add_tags=[], tag=[], tags_fetch=True, tags_in=None, tags_except=None,
+              title=None, comment=None, immutable=None, offline=False, url_redirect=False,
+              tag_redirect=False, tag_error=False, del_range=None, del_error=None):
+    argv = ['--add', (url := 'https://example.com/')] + add_tags
+    if tag:
+        argv += ['--tag'] + tag
+    if title is not None:
+        argv += ['--title', title]
+    if comment is not None:
+        argv += ['--comment', comment]
+    if immutable is not None:
+        argv += ['--immutable', str(int(immutable))]
+    if offline:
+        argv += ['--offline']
+    if url_redirect:
+        argv += ['--url-redirect']
+    if tag_redirect:
+        argv += ['--tag-redirect'] + ([] if isinstance(tag_redirect, bool) else [tag_redirect])
+    if tag_error:
+        argv += ['--tag-error'] + ([] if isinstance(tag_error, bool) else [tag_error])
+    if del_error:
+        argv += ['--del-error'] + del_range
+    with pytest.raises(SystemExit):
+        buku.main(argv)
+    network_test = url_redirect or tag_redirect or tag_error or del_error
+    fetch = not offline and (network_test or tags_fetch or title is None)
+    bdb.add_rec.assert_called_with(
+        url, title, tags_in, comment, immutable, delay_commit=False, fetch=fetch,
+        tags_fetch=tags_fetch, tags_except=tags_except, url_redirect=url_redirect,
+        tag_redirect=tag_redirect, tag_error=tag_error, del_error=del_error)
+    bdb.searchdb.assert_not_called()
+    prompt.assert_not_called()
```

Comment:
These are mostly due to initial issues I was having with `main()` (…which were basically fixed by making `sys.argv[1:]` the default value; but I left this in just to be sure)